colexec: WIP on context cancellation error

yuzefovich opened the pull request:
Release note: None

yuzefovich committed:
colexec: fix closing of parallel unordered sync inputs

The responsibility of closing all `Closer`s within the synchronizer
input's tree lies on the goroutine that's handling that input.
Previously, we were mistakenly calling `Close` after we have notified
the wait groups that the handler-goroutine is done. This can result in
races between vectorized flow cleanup and cleaning up those closers.
This is now fixed.

Release note: None

yuzefovich committed:
rowflow,colexec: make routers propagate errors to all non-closed outputs

This commit changes the way we propagate the errors in the hash router
so that the error metadata is sent on all non-closed streams.
Previously, we would be sending it over only the first non-closed stream
which could result in the processors on the same stage as that single
stream end to treat the absence of rows and errors as the input being
exhausted successfully, which is wrong because the input did encounter
an error.

The same thing has been happening in the vectorized flow, but in that
case the problem is less severe - the issue will present itself only
when we have wrapped processors (because the materializers will prevent
the propagation throughout the whole flow as described below):
In the vectorized engine we use panic-catch mechanism of error
propagation, and we end up with the following sequence of events:
1. an operator encounters an error on any node (e.g. `colBatchScan`
encounters RWUI error on a remote node). It is not an internal vectorized
error, so the operator will panic with `colexecerror.ExpectedError`.
2. the panic is caught by one of the catchers (it can be a parallel
unordered synchronizer goroutine, an outbox goroutine, a materializer,
a hash router)
3. that component will then decide how to propagate the error further:
3.1 if it is a parallel unordered synchronizer, then it will cancel all
of its inputs and will repanic
3.2 if it is an outbox, the error is sent as metadata which will be
received by an inbox which will panic with it
3.3. if it is a materializer, then it might swallow the error (this is
the reason we need for the vectorized hash router to send the error to
all of its inputs). The swallowing is acceptable if it is the root
materializer though.
3.4 if it is a hash router, it'll cancel all of its outputs and will
forward the error on each of the outputs.

Release note (bug fix): Previously, CockroachDB could return incorrect
results on query that encountered ReadWithinUncertaintyInterval error,
and this has been fixed.

yuzefovich committed:
colexec: WIP on context cancellation error

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51701)
<!-- Reviewable:end -->

yuzefovich committed:
colexec: swallow "query canceled" in synchronizer with internal cancellation

Parallel unordered synchronizer shuts down all of its input subtrees by
canceling the context which those inputs are listening on. This results
in an "internal cancellation", and the error handling is different in
such case in the synchronizer - it needs to swallow "context canceled"
errors because, well, it was the one that canceled it, so such error is
expected part of the "graceful" shutdown. However, we also need to
swallow "query canceled" errors because `CancelChecker` operator might
be listening on `ctx.Done` channel and will emit query cancellation
error when it sees that the context is canceled (so "query canceled"
error is also expected as part of the graceful shutdown). Previously, we
weren't swallowing it, and we do.

Release note: None
